rflx/model/message.py

```diff
@@ -1072,20 +1085,19 @@ def __compute_topological_sorting(self) -> Optional[Tuple[Field, ...]]:
             return None
         return tuple(f for f in result if f not in [INITIAL, FINAL])
 
-    def __compute_paths(self, final: Field) -> Set[Tuple[Link, ...]]:
-        if final == INITIAL:
-            return {()}
-        return set(
-            itertools.chain.from_iterable(
-                ((p + (l,) for p in self.__compute_paths(l.source)) for l in self.incoming(final))
-            )
-        )
+    def __compute_reverse(self) -> Dict[Field, List[Link]]:
+        result: Dict[Field, List[Link]] = {
+            f: [] for l in self.structure for f in (l.source, l.target)
+        }
+        for l in self.structure:
+            result[l.target].append(l)
+        return result
```

Comment:
It is not clear to me what `reverse` means. It looks like it is creating a dict containing all incoming edges to a field, which is the same functionality as `incoming` provides.

Reply:
It's the same information, but I though it would be more performant to not iterating over `structure` over and over again. Turns out that there is no performance gain - I'm using `incoming` now in 9f63295.

Given the surprising results when using `lru_cache` in the expressions, I think we may potentially benefit from using `lru_cache` in methods like `incoming`. As `Message` is not hashable, this is not easily possible, though (and we may loose all performance gains when this complex object is hashed...).

Reply:
> Given the surprising results when using `lru_cache` in the expressions, I think we may potentially benefit from using `lru_cache` in methods like `incoming`. As `Message` is not hashable, this is not easily possible, though (and we may loose all performance gains when this complex object is hashed...).

I think we should just try it out. As the only requirement for `__hash__` is that two equal objects must have the same hash, the implementation for `Message` can be really simple:

```Python
    def __hash__(self) -> int:
        return hash(self.identifier)
```

Or even:

```Python
    def __hash__(self) -> int:
        return 0
```

Reply:
I tried this and it doesn't improve performance (when added to certain methods, it even gets worse).